backend/admission.go

```diff
@@ -0,0 +1,126 @@
+package backend
+
+import (
+	"context"
+
+	"github.com/grafana/grafana-plugin-sdk-go/genproto/pluginv2"
+)
+
+// AdmissionHandler is an EXPERIMENTAL service that allows checking objects before they are saved
+// This is modeled after the kubernetes model for admission controllers
+// Since grafana 11.1, this feature is under active development and will continue to evolve in 2024
+// This may also be replaced with a more native kubernetes solution that does not work with existing tooling
+type AdmissionHandler interface {
```

Comment:
Nit. One thing that would be nice/useful is if a plugin doesn't have to implement all of these and return not implemented/custom errors - better if SDK can provide good defaults if not implemented. For example always return allowed = 1 for validation if not implemented. Then there's no risk of calling validate when creating/updating a datasource and not having to check for unimplemented errors.

```go
type ValidateAdmissionHandler interface {
	// ValidateAdmission is a simple yes|no check if an object can be saved
	ValidateAdmission(context.Context, *AdmissionRequest) (*ValidationResponse, error)
}

// ValidateAdmissionFunc implements ValidateAdmissionHandler.
type ValidateAdmissionFunc func(context.Context, *AdmissionRequest) (*ValidationResponse, error)

func (f ValidateAdmissionFunc) ValidateAdmission(ctx context.Context, req *AdmissionRequest) (*ValidationResponse, error) {
	return f(ctx, req)
}

type MutateAdmissionHandler interface {
	// MutateAdmission converts the input into an object that can be saved, or rejects the request
	MutateAdmission(context.Context, *AdmissionRequest) (*MutationResponse, error)
}

// MutateAdmissionFunc implements MutateAdmissionHandler.
type MutateAdmissionFunc func(context.Context, *AdmissionRequest) (*MutationResponse, error)

func (f MutateAdmissionFunc) MutateAdmission(ctx context.Context, req *AdmissionRequest) (*MutationResponse, error) {
	return f(ctx, req)
}

type ConvertObjectHandler interface {
	// ConvertObject is called to covert objects between different versions
	ConvertObject(context.Context, *ConversionRequest) (*ConversionResponse, error)
}

// ConvertObjectFunc implements ConvertObjectHandler.
type ConvertObjectFunc func(context.Context, *ConversionRequest) (*ConversionResponse, error)

func (f ConvertObjectFunc) ConvertObject(ctx context.Context, req *ConversionRequest) (*ConversionResponse, error) {
	return f(ctx, req)
}

type AdmissionHandler interface {
	ValidateAdmissionHandler
	MutateAdmissionHandler
	ConvertObjectHandler
}

// to be used in ManageOpts
type AdmissionHandlers struct {
  ValidateAdmission ValidateAdmissionHandler
  MutateAdmission MutateAdmissionHandler
  ConvertObject ConvertObjectHandler
}

Reply:
interesting -- lets find an approach in a follow up PR.  Right now the clients will implement everything so you do not know if an implementation exists or not.  The Conversion service will only be needed if multiple versions exist -- so that is a good canidate

Reply:
Sounds good 👍

Reply:
In k8s land you subscribe to webhooks which clearly indicates which one of validate/mutate your interest in so something similar from plugin dev perspective was what I had in mind